### PR TITLE
Problem: Missing IBC denom trace to hashed denom API

### LINKED ISSUE #490: Problem: Missing IBC denom trace to hashed denom API

### PR DESCRIPTION
**Pull request by ysong42**
Solution: fixed #490 

**This one is depended on PR #511 (ibc channel debug projection), please review that one first** 🙏 

### Changes

- Introduced a new table, which contain two columns `denom` and `hash`
- Added a new API to fetch all denom-hash mappings. `/ibc/denom-hash-mappings`

### Example API Response

```
{
    "result": [
        {
            "denom": "transfer/channel-0/solotoken",
            "hash": "1A35E932DCE61466ED9F72D0B436628C388FB1BC60CB23A055039B7DE54883CC"
        },
        {
            "denom": "transfer/channel-1/USDC",
            "hash": "C59CA0D4F324B119C79782232700A6920B2EACA776558D7319CFC36DEE5672BF"
        },
        {
            "denom": "transfer/channel-2/USDC",
            "hash": "21E9F91EE9AC1D7E2C3D002F4AA73CAAB61801B055FB6721D36F16BB890BE0A2"
        },
        {
            "denom": "transfer/channel-1/TEST",
            "hash": "96D929A2DDBAB720DA3AB084914A64D40920DCF8CCD06B32966FBAF3AE75BD24"
        },
        {
            "denom": "transfer/channel-2/TEST",
            "hash": "D6F89C93C2DA3753E81650E96153A2C3E6BCC781E42638AE7BE2635476E2B6BB"
        },
        {
            "denom": "transfer/channel-2/OTTER",
            "hash": "C63FD640F6213F1D23C2D28E73A7289E9524F8B6156107E40186834E27B1D3CE"
        }
    ]
}
```

### CODE REVIEW

File: migrations/20210902150815_view_ibc_denom_hash_mapping.up.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE view_ibc_denom_hash_mapping (
+    id BIGSERIAL,
+    denom VARCHAR NOT NULL UNIQUE,
+    hash VARCHAR NOT NULL,
```
Comment (davcrypto) [inline migrations/20210902150815_view_ibc_denom_hash_mapping.up.sql:4]: Is the `hash`  need to be index as well?

Comment (ysong42) [reply migrations/20210902150815_view_ibc_denom_hash_mapping.up.sql:4]: Thanks for pointing this out! Yes agree, later we may have requirement to return the denom of a given hash.

**Review by davcrypto — APPROVED** [review]

File: projection/ibc_channel/view/ibc_denom_hash_mapping.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	rdb *rdb.Handle
+}
+
+func NewIBCDenomHashMapping(handle *rdb.Handle) *IBCDenomHashMapping {
+	return &IBCDenomHashMapping{
+		handle,
+	}
+}
+
+func (ibcDenomHashMappingView *IBCDenomHashMapping) IfDenomExist(denom string) (bool, error) {
```
Comment (calvinaco) [inline projection/ibc_channel/view/ibc_denom_hash_mapping.go:20]: Consider using `UPSERT`
https://stackoverflow.com/questions/4069718/postgres-insert-if-does-not-exist-already

**Review by calvinaco — APPROVED** [review]
Logic looks good to me

File: projection/ibc_channel/view/ibc_denom_hash_mapping.go (hunk 2 above)
Comment (ysong42) [reply projection/ibc_channel/view/ibc_denom_hash_mapping.go:20]: I think `UPSERT` is a more expensive operation compare to `EXISTS`. 

As we only need to insert the record for once, and the record is not supposed to be updated later.

**Pull request merged**